File: go/leet_code/79.go
// Package leet_code provides solutions for LeetCode problems.
//
// This file provides a solution for the problem of determining if a word exists in a 2D grid of characters.
// The word can be constructed from letters of sequentially adjacent cells, where "adjacent" cells are horizontally or vertically neighboring.
// The same letter cell may not be used more than once.
package leet_code

// dfs is a helper function that performs a depth-first search on the board starting from the cell at (i, j).
//
// It accepts five parameters:
// - board: a 2D byte array representing the board.
// - i, j: integers representing the starting cell coordinates.
// - word: a string representing the word to be found.
// - k: an integer representing the current character of the word.
//
// The function returns a boolean indicating whether the word can be found starting from the cell at (i, j).
//
// The function checks the current cell and recursively checks its adjacent cells.
//
// Time complexity analysis:
// - Best-case: O(1), when the word is found in the first cell.
// - Worst-case: O(m*n*4^k), where m and n are the dimensions of the board and k is the length of the word.
// - Average-case: O(m*n*4^k), as in the worst case.
func dfs(board [][]byte, i, j int, word string, k int) bool {
	if i < 0 || i >= len(board) || j < 0 || j >= len(board[i]) {
		return false
	}

	if board[i][j] != word[k] {
		return false
	}

	if k == len(word)-1 {
		return true
	}

	tmp := board[i][j]
	board[i][j] = '.'

	if dfs(board, i-1, j, word, k+1) || dfs(board, i+1, j, word, k+1) || dfs(board, i, j-1, word, k+1) || dfs(board, i, j+1, word, k+1) {
		return true
	}

	board[i][j] = tmp

	return false
}

// exist determines if a word exists in a 2D grid of characters.
//
// It accepts two parameters:
// - board: a 2D byte array representing the board.
// - word: a string representing the word to be found.
//
// The function returns a boolean indicating whether the word exists in the board.
//
// The function iterates over the board and for each cell, it calls the dfs function to check if the word exists starting from that cell.
//
// Time complexity analysis:
// - Best-case: O(1), when the word is found in the first cell.
// - Worst-case: O(m*n*4^k), where m and n are the dimensions of the board and k is the length of the word.
// - Average-case: O(m*n*4^k), as in the worst case.
func exist(board [][]byte, word string) bool {
	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[i]); j++ {
			if dfs(board, i, j, word, 0) {
				return true
			}
		}
	}

	return false
}

File: ruby/leet_code/79.rb
# Function to check if a word exists in a 2D board
#
# This function checks if a given word can be formed in the board by sequentially connecting cells horizontally or vertically.
# Each cell in the board contains a single letter, and the same letter cell may not be used more than once.
#
# @param board [Array<Array<String>>] The 2D board of letters
# @param word [String] The word to be checked
# @return [Boolean] True if the word exists in the board, false otherwise
#
# @example
#   exist([['A','B','C','E'],['S','F','C','S'],['A','D','E','E']], "ABCCED") #=> true
#
def exist(board, word)
  if word.match(/^#{word[0]}+/)[0].size > word.match(/#{word[-1]}+$/)[0].size
    word.reverse!
  end
  board.length.times do |i|
    board[0].length.times do |j|
      if board[i][j] == word[0]
        res = search(board, i, j, word)
        return true if res
      end
    end
  end
  false
end

# Function to search for a word in the board starting from a given cell
#
# @param board [Array<Array<String>>] The 2D board of letters
# @param i [Integer] The row index of the starting cell
# @param j [Integer] The column index of the starting cell
# @param word [String] The word to be searched
# @param idx [Integer] The index of the current character in the word
# @return [Boolean] True if the word is found, false otherwise
#
def search(board, i, j, word, idx = 0)
  if idx == word.length - 1 && board[i][j] == word[idx]
    true
  elsif board[i][j] != word[idx]
    return false
  else
    temp = board[i][j]
    board[i][j] = "*"
    [[-1, 0], [1, 0], [0, 1], [0, -1]].each do |off|
      ip, jp = i + off[0], j + off[1]
      if in_bounds?(ip, jp, board)
        res = search(board, ip, jp, word, idx + 1)
        return res if res
      end
    end
    board[i][j] = temp
    false
  end
end

# Function to check if a cell is within the bounds of the board
#
# @param i [Integer] The row index of the cell
# @param j [Integer] The column index of the cell
# @param board [Array<Array<String>>] The 2D board of letters
# @return [Boolean] True if the cell is within the bounds of the board, false otherwise
#
def in_bounds?(i, j, board)
  i >= 0 && j >= 0 && i < board.length && j < board[0].length
end